skiko/src/jsMain/kotlin/org/jetbrains/skiko/wasm/api/CanvasRenderer.kt

```diff
@@ -0,0 +1,48 @@
+package org.jetbrains.skiko.wasm.api
+
+import kotlinx.browser.window
+import org.jetbrains.skia.BackendRenderTarget
+import org.jetbrains.skia.Canvas
+import org.jetbrains.skia.ColorSpace
+import org.jetbrains.skia.DirectContext
+import org.jetbrains.skia.Surface
+import org.jetbrains.skia.SurfaceColorFormat
+import org.jetbrains.skia.SurfaceOrigin
+import org.jetbrains.skiko.wasm.CreateWebGLContext
+import org.jetbrains.skiko.wasm.GL
+import org.w3c.dom.HTMLCanvasElement
+
+
+fun Surface.Companion.createFromGL(width: Int, height: Int) = makeFromBackendRenderTarget(
+    DirectContext.makeGL(),
+    BackendRenderTarget.makeGL(
+        width, height, 1, 8, 0, 0x8058
+    ),
+    SurfaceOrigin.BOTTOM_LEFT,
+    SurfaceColorFormat.RGBA_8888,
+    ColorSpace.sRGB
+)
+
+abstract class CanvasRenderer constructor(htmlCanvas: HTMLCanvasElement, val width: Int, val height: Int) {
+    constructor(canvas: HTMLCanvasElement): this(canvas, canvas.width, canvas.height)
+    private val contextPointer = CreateWebGLContext(htmlCanvas)
+
+    init {
+        GL.makeContextCurrent(contextPointer)
+    }
+
+    val surface = Surface.createFromGL(width, height)
+    val canvas: Canvas
+        get() = surface.canvas
+
+    abstract fun drawFrame(currentTimestamp: Double)
```

Comment:
Maybe make `
class CanvasRenderer constructor(htmlCanvas: HTMLCanvasElement, val width: Int, val height: Int, val drawFrame: (Double) -> Unit)` instead to avoid unneeded subclassing?

Reply:
This would be  a quite javascript thing to do, however my arguments for not doing so are following:
1) In future we might need more than one method to implement (for instance, needsRedraw)
2) By the very nature of  usage of such renderer there most likely be a data which is not supposed to be calculated in onDraw  - that said, in real-life scenarious there still will be some class/object to work with.